hushline/crypto/secrets_manager.py

```diff
@@ -0,0 +1,113 @@
+"""
+A class definition to simplify the careful handling of cryptographic secrets.
+"""
+
+import warnings
+from base64 import b64decode, urlsafe_b64encode
+from hashlib import shake_256
+from pathlib import Path
+from secrets import token_bytes
+
+from aiootp.generics.canon import canonical_pack, fullblock_ljust
+from cryptography.fernet import Fernet, InvalidToken
+from passlib.hash import argon2
+
+__all__ = ["InvalidToken", "SecretsManager"]
+
+
+def truncated_b64decode(value: bytearray) -> bytearray:
+    if bool(remainder := len(value) % 4):
+        value.extend((4 - remainder) * b"=")
+    return bytearray(b64decode(value))
+
+
+class SecretsManager:
+    __slots__ = ("_secret_salt_filename", "_secret_salt_length", "_memory_cost", "_kdf")
+
+    _APP_SECRETS_DIRECTORY: Path = Path("hushline/crypto/_app_secrets")
+    _KDF_BLOCKSIZE: int = shake_256().block_size
+    _MEMORY_COST_IN_KiB: int = 128 * 1024
+    _SECRET_SALT_FILENAME: str = "_device_salt.txt"
+    _SECRET_SALT_LENGTH: int = 32
+
+    def __init__(
+        self,
+        admin_secret: bytearray,
+        *,
+        memory_cost: int | None = None,
+        secret_salt_filename: str | None = None,
+        secret_salt_length: int | None = None,
+    ) -> None:
+        self._APP_SECRETS_DIRECTORY.mkdir(parents=True, exist_ok=True)
+        self._memory_cost = self._MEMORY_COST_IN_KiB if memory_cost is None else memory_cost
+        self._secret_salt_filename = (
+            self._SECRET_SALT_FILENAME if secret_salt_filename is None else secret_salt_filename
+        )
+        self._secret_salt_length = (
+            self._SECRET_SALT_LENGTH if secret_salt_length is None else secret_salt_length
+        )
+        self._prepare_key_derivation_object(admin_secret)
+
+    def _summon_device_salt(self) -> bytearray:
+        salt_path = self._APP_SECRETS_DIRECTORY / self._secret_salt_filename
+
+        if not salt_path.is_file():
+            salt_path.write_bytes(salt := token_bytes(self._secret_salt_length))
+        else:
+            salt_path.chmod(0o600)
+            if len(salt := salt_path.read_bytes()) != self._secret_salt_length:
+                warnings.warn("The secret salt length doesn't match its declaration.", stacklevel=2)
+
+        salt_path.chmod(0o000)
+        return bytearray(salt)
+
+    def _prepare_key_derivation_object(self, admin_secret: bytearray) -> None:
+        salt = self._summon_device_salt()
+        hashed_secret_with_metadata = bytearray(
+            argon2
+            .using(salt=bytes(salt), digest_size=self._KDF_BLOCKSIZE, memory_cost=self._memory_cost)
+            .hash(bytes(admin_secret)),
+            encoding="utf-8",
+        )
+        hashed_secret = truncated_b64decode(hashed_secret_with_metadata.split(b"$")[-1])
```

Comment:
what's the purpose of this truncated `hashed_secret` value vs initializing the hash with the hashed secret+metadata?

Reply:
This is a fine detail regarding how to turn a shake_256 object into a keyed permutation. About a year ago I wrote about a more in-depth security [analysis](https://crypto.stackexchange.com/a/107513) of keyed SHA-3 objects. But essentially, the reason is because it's easier to make security arguments when using [_uniformly random_](https://en.wikipedia.org/wiki/Discrete_uniform_distribution) keys. But argon2 output looks like this:

```bash
'$argon2id$v=19$m=65536,t=3,p=4$ACAkhBCCkLL2XotxrnXOuQ$AW3E2OwuNEId...'
```

This metadata, and the encoding, are not uniformly random. So, only the raw, random hash part of the output is used as the key to initialize the keyed internal state. After it's initialized, the whole argon2 output is ingested as a commitment to the encoded original, and then padded to a multiple of the internal block-size.

https://github.com/scidsg/hushline/blob/160fa0128723c6296148d41c363d56bb66bd5aec/hushline/crypto/secrets_manager.py#L74


Here's a bit of a visual to see what's going on:


```bash
'''
# the initial shake_256 state is 200 bytes of zeros
"00000000000000000000000000000000000000000000000000"
                 ⊕
# the raw hashed secret is exactly the size of the internal block-size
"hashedsecrethashedsecrethashedsecr"
                 =
# the raw hashed secret gets XORd into the state
"hashedsecrethashedsecrethashedsecr0000000000000000"

# then, scrambling occurs after a block-size number of bytes are ingested
π("hashedsecrethashedsecrethashedsecr0000000000000000")
                 =
# after the scrambling, the whole internal state is randomized
"0ih7zekm1l5cea6i6yy5bx54bk9pd4e9k3u4zpgbpah675dd5m"
'''
```




---






Another visual of the internal [SHA-3](https://en.wikipedia.org/wiki/SHA-3) "sponge":

[![Sponge-Construction-Keccak-000](https://github.com/user-attachments/assets/56d4c4a5-197a-4803-84b1-7a27167f69ee)](https://keccak.team/sponge_duplex.html)

Reply:
This seems like a great reason, let's just add a comment to that effect, something like "The initial key of the shake_256() should be uniformly random, so strip the metadata and encoding from the hashed secret to use as our initial key"

Reply:
Addressed by (https://github.com/scidsg/hushline/pull/411/commits/4dd848812843242de2959eed0afdec89b6cc16a5)